Windows support

hasufell opened the pull request.

hsyl20 committed:
Fix PIC jump tables on Windows (#24016)

Avoid overflows in jump tables by using a base label closer to the jump
targets. See added Note [Jump tables]

hasufell committed:
Windows support

hasufell commented on cabal.project.stage2:
```diff
@@ -64,7 +64,7 @@ packages:
 
   libraries/haskeline
-  libraries/terminfo
+--  libraries/terminfo
```
only on windoof

hasufell committed:
ghc-toolchain hack adding ways

angerman commented on testsuite/driver/testlib.py:
```diff
@@ -3266,7 +3266,7 @@ async def runCmd(cmd: str,
         # to invoke the Bourne shell
 
         proc = await asyncio.create_subprocess_exec(timeout_prog, timeout, cmd,
-                                                   stdin=stdin_file,
+                                                   stdin=stdin_file if stdin_file else asyncio.subprocess.PIPE,
                                                    stdout=asyncio.subprocess.PIPE,
```
This feels like a prime example of something we should upstream?

hasufell replied:
You sure you want me to spend 3 weeks on that?

angerman commented on testsuite/tests/ghc-api/all.T:
```diff
@@ -61,7 +61,7 @@ test('T19156', [ extra_run_opts('"' + config.libdir + '"')
                ['-package ghc'])
-test('T20757', [unless(opsys('mingw32'), skip), exit_code(1), normalise_version('ghc')],
+test('T20757', [unless(opsys('mingw32'), skip), exit_code(0), normalise_version('ghc')],
                compile_and_run,
```
exit code changed all of a sudden?

hasufell replied:
this has been removed

angerman commented on testsuite/tests/rts/linker/T20494-obj.c:
```diff
@@ -3,9 +3,13 @@
 #define DESTRUCTOR(prio)  __attribute__((destructor(prio)))
+#if defined(_WIN32)
+#define PRINT(str) printf(str); fflush(stdout)
+#else
 // don't use "stdout" variable here as it is not properly defined when loading
 // this object in a statically linked GHC.
 #define PRINT(str) dprintf(1,str); fsync(1)
+#endif
```
We should probably create an issue and fix this in ghc as well. why on earth is stdout the wrong one here? That's really unexpected and confusing.

angerman replied:
@copilot create an issue for this.

hasufell replied:
This depends on a change we did downstream: https://github.com/stable-haskell/ghc/commit/86e4c0afe2c2e36354ef04b81706f24052654ff2#diff-9926f97216db79147f93e10d0afa4d8d59b218723f7f03cbeb1aac25d231ad92R6

So I don't think there's anything to upstream.

angerman replied:
I think the downstream "fix" papers over an upstream issue around handling stdout in the linker properly.

angerman commented on testsuite/tests/rts/T23142.hs:
```diff
@@ -1,5 +1,5 @@
 {-# LANGUAGE UnboxedTuples, MagicHash #-}
-module T23142 where
+module Main where
```
Main?

angerman replied:
Why is this called Main now? I think this change needs to be reverted?
```suggestion
module T23142 where
```

hasufell replied:
Because it failed without Main :hankey:

angerman replied:
You are kidding, right? Right?!

angerman commented on cabal.project.stage2:
```diff
@@ -87,12 +90,15 @@ constraints:
   executable-profiling: False
   executable-static: False
 
+if !os(windows)
+  package *
+    library-for-ghci: True
```
I hope we can get rid of this outright soon 😅 
If we go with dynamic by default, we probably can.
But only on platforms where dynamic by default works.
The altenrative is that we need to rework the in-memory linker to be more powerful.

Or, what probably is even a better idea, to delay this, and have GHC _ondeman_ ld -r (on platforms that suppor thtat) create the ghci objects from archives.

angerman commented on cabal.project.stage2:
```diff
@@ -142,6 +148,23 @@ if os(freebsd)
+    ghc-options: "-optc-DBuildVendor=\"FIXME\""
+    ghc-options: "-optc-DGhcUnregisterised=\"FIXME\""
+    ghc-options: "-optc-DTablesNextToCode=\"FIXME\""
+    ghc-options: "-optc-DFS_NAMESPACE=rts"
```
We should really figure out which of these we need.

angerman commented on Makefile:
```diff
@@ -99,10 +99,22 @@ ROOT_DIR := $(patsubst %/,%,$(dir $(realpath $(lastword $(MAKEFILE_LIST)))))
+ifeq ($(OS),Windows_NT)
+CC := x86_64-w64-mingw32-clang.exe
+CXX := x86_64-w64-mingw32-clang++.exe
+CC_LINK_OPT := -Wl,CRT_fp8.o
+LD := ld.lld.exe
+CYGPATH = cygpath --unix -f -
+CYGPATH_MIXED = cygpath --mixed -f -
```
We should document what especially the CC_LINK_OPT, CGPATH, CYGPATH mixed env vars do?

angerman replied:
@copilot Can you provide a suggested change for the documentation?

angerman commented on Makefile:
```diff
@@ -99,10 +99,22 @@ ROOT_DIR := $(patsubst %/,%,$(dir $(realpath $(lastword $(MAKEFILE_LIST)))))
+else
+CYGPATH_MIXED = cat
+CYGPATH = cat
++CC_LINK_OPT ?=
```
```suggestion
CC_LINK_OPT ?=
```
?

angerman commented on Makefile:
```diff
@@ -167,6 +179,7 @@ TARGET_PLATFORM = $(call GHC_INFO,target platform string)
 TARGET_ARCH     = $(call GHC_INFO,target arch)
 TARGET_OS       = $(call GHC_INFO,target os)
 TARGET_TRIPLE   = $(call GHC_INFO,Target platform)
+GHC_LIBDIR      = $(call GHC_INFO,LibDir)
```
Can't GHC figure out it's own LIBDIR? 🤔  I thought that was implemented on windows, macos, and linux?

hasufell replied:
This is for the `cabal.project.stage1.local` target:

```
cabal.project.stage1.local: cabal.project.stage1
ifeq ($(OS),Windows_NT)
	echo "extra-prog-path: $(shell echo '$(GHC_LIBDIR)' | $(CYGPATH_MIXED))/../mingw/bin" > $@
else
	echo "" > $@
endif
```

angerman reviewed (commented):
Looks pretty impressive! nice work @hasufell!

hasufell committed:
rts: Fix object file format detection in loadArchive

Commit 76d1041dfa4b96108cfdd22b07f2b3feb424dcbe seems to
have introduced this bug, ultimately leading to failure of
test T11788. I can only theorize that this test isn't run
in upstream's CI, because they don't build a static GHC.

The culprit is that we go through the thin archive, trying
to follow the members on the filesystem, but don't
re-identify the new object format of the member. This pins
`object_fmt` to `NotObject` from the thin archive.

Thanks to @angerman for spotting this.

angerman committed:
rts: Initialize gc_thread timing fields in new_gc_thread

The gc_thread timing fields (gc_start_cpu, gc_end_cpu, gc_start_elapsed,
gc_end_elapsed, gc_sync_start_elapsed) were not being initialized when
gc_threads were allocated. Since gc_threads are allocated with
stgMallocAlignedBytes (which doesn't zero memory), these fields contained
garbage values.

The initialization must be in new_gc_thread(), not init_gc_thread(),
because:

1. new_gc_thread() is called once when a gc_thread is first allocated
2. init_gc_thread() is called at the START of each GC cycle
3. stat_startGC() sets the timing fields BEFORE init_gc_thread() is called
4. If we initialize in init_gc_thread(), we would reset the values that
   stat_startGC() just set, breaking the timing calculations

The garbage values caused wild statistics like:
  gc_elapsed_ns=50426020081527 (14 hours of supposed GC time!)
  exit_elapsed_ns=18446741672370457118 (~= -1.3 billion as unsigned)

These were being accumulated into stats and causing all productivity
calculations to fail with massively negative values.

angerman committed:
rts: Add stgCallocAlignedBytes with zeroing semantics

Introduce stgCallocAlignedBytes as a zeroing aligned allocator, replacing
stgMallocAlignedBytes. This allows removing ~40 lines of redundant zero/NULL
initializations in new_gc_thread() and initCapability().

Changes:
- Rename stgMallocAlignedBytes to stgCallocAlignedBytes and add memset(0)
- Add deprecated stgMallocAlignedBytes wrapper for backwards compatibility
- Update call sites in GC.c and Capability.c to use stgCallocAlignedBytes
- Remove redundant zero/NULL/false initializations from:
  - new_gc_thread(): timing fields, free_blocks, gc_count, workspace fields
  - initCapability(): most boolean/numeric/pointer fields

The zeroing overhead is negligible (startup-time allocation, ~500-1000 bytes)
while the benefits include:
- Cleaner code with only non-zero initializations remaining
- Safer: new struct fields automatically start at zero
- Catches uninitialized memory bugs (was causing garbage timing values)

hasufell committed:
Bump mac x86 runners to macOS-15-intel

See https://github.com/actions/runner-images/issues/13046

angerman commented on utils/ghc-toolchain/exe/Main.hs:
```diff
@@ -607,7 +607,7 @@ targetToSettings tgt@Target{..} =
   , ("Use interpreter", yesNo (targetSupportsInterpreter tgt))
   , ("Support SMP", yesNo (targetSupportsSMP tgt))
-  , ("RTS ways", "v") -- FIXME: should be a property of the RTS, not of the target
+  , ("RTS ways", "v thr debug thr_debug") -- FIXME: should be a property of the RTS, not of the target
```
@copilot Open an issue for this. So we track this defect.

hasufell replied:
The reason for this is solely to trick the test suite into running all the tests. It has no other impact.

angerman commented on .github/workflows/reusable-release.yml:
```diff
@@ -103,18 +111,22 @@ jobs:
+          retry_on: error
+          command: |
+            set -eux
+            curl -O -L https://ftp.gnu.org/gnu/make/make-${{ env.MAKE_VERSION }}.tar.gz
```
For some weird reason this keeps failing every now and then. Should we cache this on a dl.stable-haskell.org/.../?

hasufell replied:
No opinion

angerman commented on .github/workflows/reusable-release.yml:
```diff
@@ -521,33 +574,46 @@ jobs:
+          PATH=$(pwd)/_build/bindist/bin:$PATH ${{ env.MAKE }} -C _build/bindist/testsuite
         env:
           THREADS: 4
+          TEST_HC: ${{ steps.pwd_output.outputs.pwd }}/_build/bindist/bin/ghc
+          GHC_PKG: ${{ steps.pwd_output.outputs.pwd }}/_build/bindist/bin/ghc-pkg
+          HP2PS_ABS: ${{ steps.pwd_output.outputs.pwd }}/_build/bindist/bin/hp2ps
+          HPC: ${{ steps.pwd_output.outputs.pwd }}/_build/bindist/bin/hpc
+          RUNGHC: ${{ steps.pwd_output.outputs.pwd }}/_build/bindist/bin/runghc
```
Don't we want to be explicit, and shouldn't these have `.exe` suffixes on windows?

angerman replied:
Something like `${{ runner.os == "Windows" && ".exe" || "" }}` should work? @copilot, will that work for the env section here?

hasufell replied:
Doesn't need suffixes... this is run via the makefile/in the shell

angerman commented on Makefile:
```diff
@@ -155,8 +170,8 @@ GHC ?= $(GHC0)
 
-GHC1 = _build/stage1/bin/ghc
-GHC2 = _build/stage2/bin/ghc
+GHC1 = _build/stage1/bin/ghc$(EXE_EXT)
+GHC2 = _build/stage2/bin/ghc$(EXE_EXT)
```
Here we _do_ use the `$(EXE_EXT)`; I think we should use the same for the test above.

angerman commented on Makefile:
```diff
@@ -234,26 +250,31 @@ STAGE1_TARGETS := $(STAGE_UTIL_TARGETS) ghc-bin:ghc ghc-toolchain-bin:ghc-toolch
 STAGE2_TARGETS := \
 	ghc-bin:ghc
 
+# we need to build these before all else
```
Why do we need to build these before all else, we should document that.

Ultimately we only want to ship one (threaded-nodebug) by default, and then have the others be lazy installable later on as needed. @copilot create an issue for this, if none exists yet.

hasufell replied:
Because the dependencies aren't tracked properly and the build fails, because some required rts is not available yet.

angerman replied:
At what point? Where is it missing?

What are we trying to build that needs a differnt rts? (Outside of the testsuite?)

angerman commented on Makefile:
```diff
@@ -263,8 +284,12 @@ STAGE2_UTIL_TARGETS := \
 	system-cxx-std-lib:system-cxx-std-lib \
-	terminfo:terminfo \
-	xhtml:xhtml
+	xhtml:xhtml \
+	haddock:haddock
```
These are only needed for the test-suite right?

hasufell replied:
haddock should be part of the distribution

angerman replied:
Why? Why can't haddock be installed on demand from hackahe?

angerman commented on Makefile:
```diff
@@ -561,39 +588,48 @@ endef
 	@mkdir -p _build/stage0/bin _build/logs
 	cabal build $(BUILD_ARGS) cabal-install:exe:cabal
-	cp -rfp $(shell cabal list-bin -v0 $(BUILD_ARGS) cabal-install:exe:cabal) _build/stage0/bin/cabal
+	cp -rfp $(shell cabal list-bin -v0 $(BUILD_ARGS) cabal-install:exe:cabal | $(CYGPATH)) $@
```
Do we still need this after @andreabedini's change to cabal to install stuff into bin anyway?

angerman commented on Makefile:
```diff
@@ -561,39 +588,48 @@ endef
 
+.PHONY: cabal.project.stage1.local
+
+cabal.project.stage1.local: cabal.project.stage1
+ifeq ($(OS),Windows_NT)
+	echo "extra-prog-path: $(shell echo '$(GHC_LIBDIR)' | $(CYGPATH_MIXED))/../mingw/bin" > $@
+else
+	echo "" > $@
+endif
```
Why is this necessary? We don't have the default tools on-path by default?

hasufell replied:
Because if we specify `CC := x86_64-w64-mingw32-clang.exe` and then pass it to ghc-toolchain, it will pick the system one and start mixing toolchains. Only *GHC* knows where the internal mingw folder is... cabal doesn't. This causes frequent issues outside of stable-haskell as well.

angerman replied:
We should document this insight as a comment here. Also, will this get easier once we don't shit our own msys2 copy anymore?

angerman commented on Makefile:
```diff
@@ -623,43 +659,57 @@ stage1: $(addprefix _build/stage1/bin/,$(STAGE1_EXECUTABLES)) _build/stage1/lib/
 		$(CABAL_BUILD) --ghc-options="-ghcversion-file=$(abspath ./rts/include/ghcversion.h)" -W $(GHC0) $(STAGE2_TARGETS)
 	@echo "::endgroup::"
 
+.PHONY: stage2-rts
+stage2-rts: private STAGE=stage2
+stage2-rts: private GHC=$(realpath _build/stage1/bin/ghc$(EXE_EXT))
+stage2-rts: private TARGET_PLATFORM=
+stage2-rts: $(CABAL) stage1 cabal.project.stage2
+	@echo "::group::Building stage2 RTSes..."
+	# Force cabal to replan
+	rm -rf _build/stage2/cache
+	GHC=$(GHC) HADRIAN_SETTINGS='$(HADRIAN_SETTINGS)' \
+		PATH='$(PWD)/_build/stage1/bin:$(PATH)' \
+		$(CABAL_BUILD) --ghc-options="-ghcversion-file=$(abspath ./rts/include/ghcversion.h)" -W $(GHC0) $(STAGE2_UTIL_RTS)
+	@echo "::endgroup::"
```
We should document why this target is needed. This is only here so we build _all_ rts's for the test-suite?

`ghc-bin` depends only on the specific one needed for its own linking, iirc. (based on whether or not it's compiled with threading, or debug).

I still think we have the proper rts tracking for the ghc we build, but this might be required for the test-suite, as we don't want/can't lazy install stuff during the test-suite? (And the whole rts lazy install stuff is still WIP).

hasufell replied:
No, it also caused build failures.

angerman commented on cabal.project.rts:
```diff
@@ -44,6 +44,13 @@ if os(wasi)
+    ghc-options: "-optc-DHostArch=\"x86_64\""
+    ghc-options: "-optc-DHostOS=\"mingw32\""
+    ghc-options: "-optc-DHostPlatform=\"x86_64-unknown-mingw32\""
+    ghc-options: "-optc-DHostVendor=\"unknown\""
```
technically this would be msys2, I guess.

angerman commented on cabal.project.rts:
```diff
@@ -44,6 +44,13 @@ if os(wasi)
+if os(windows)
+  package rts
+    ghc-options: "-optc-DHostArch=\"x86_64\""
+    ghc-options: "-optc-DHostOS=\"mingw32\""
+    ghc-options: "-optc-DHostPlatform=\"x86_64-unknown-mingw32\""
```
I guess what we are really targeting here is mingw32-ucrt64; instead of mingw32-mscvcrt; no idea how to properly encode this. We should leave it for now. 

@copilot what's your take?

angerman commented on cabal.project.stage2:
```diff
@@ -99,14 +102,24 @@ constraints:
 --
 
+package haddock-api
+  flags: +in-ghc-tree
```
This flag is so stupid 🤦

angerman commented on cabal.project.stage2:
```diff
@@ -45,6 +45,9 @@ packages:
   utils/hp2ps
   utils/runghc
   utils/unlit
+  utils/haddock
+  utils/haddock/haddock-api
+  utils/haddock/haddock-library
```
Again we only need haddock for the test-suite, right?

hasufell replied:
Haddock should be part of the bindist.

angerman replied:
Why?

angerman commented on cabal.project.stage2:
```diff
@@ -148,7 +161,7 @@ package ghc
 
 package ghc-bin
-  flags: +internal-interpreter -threaded
+  flags: +internal-interpreter
```
Can we be explict about threaded? +threaded?

angerman commented on compiler/GHC/SysTools/BaseDir.hs:
```diff
@@ -169,21 +163,4 @@ findToolDir
-              oneLevel <- doesDirectoryExist try
-              if oneLevel
-                then return (Just path)
-                else go (k+1) (path </> "..") tried'
-findToolDir True _ = return Nothing
-#else
 findToolDir _ _ = return Nothing
```
I guess this is now a dead function? Should we deprecate/delete it?

hasufell replied:
No opinion

angerman commented on libraries/ghc-internal/config.guess:
Why do we inline the .guess and .sub now? Not that I'm fundamentally opposed, it's just not obvious to me. I remember having to do this when apple's arm machines were new to support aarch64-darwin, but this seems to have a different rational?

hasufell replied:
It caused build failures on two platforms. I have no idea why.

angerman commented on rts/RtsStartup.c:
```diff
@@ -91,15 +90,7 @@ static void flushStdHandles(void);
 static void
 x86_init_fpu ( void )
 {
-#if defined(mingw32_HOST_OS) && defined(x86_64_HOST_ARCH) && !X86_INIT_FPU
-    /* Mingw-w64 does a stupid thing. They set the FPU precision to extended mode by default.
-    The reasoning is that it's for compatibility with GNU Linux ported libraries. However the
-    problem is this is incompatible with the standard Windows double precision mode.  In fact,
-    if we create a new OS thread then Windows will reset the FPU to double precision mode.
-    So we end up with a weird state where the main thread by default has a different precision
-    than any child threads. */
-    fesetenv(FE_PC53_ENV);
-#elif X86_INIT_FPU
```
Don't we still need to somehow call `fpreset`?

hasufell replied:
no

angerman replied:
Is this documented anywhere? I know we don't the fp8 loading now, but I didn't know we don't need to call fpresrt anymore? I thought the fp8 object just brought the right fpreset into scope

angerman commented on testsuite/tests/codeGen/should_run/T25374/all.T:
```diff
@@ -1,3 +1,3 @@
 # This shouldn't crash the disassembler
-test('T25374', [extra_hc_opts('+RTS -Di -RTS'), ignore_stderr, req_target_debug_rts], ghci_script, [''])
+test('T25374', [fragile(0), extra_hc_opts('+RTS -Di -RTS'), ignore_stderr, req_target_debug_rts], ghci_script, [''])
```
So this is _just_ fragile, and we don't have an issue for it being fragile?

hasufell replied:
Afair this failed only on some platforms and I simply disabled it for that reason.

angerman commented on .github/workflows/reusable-release.yml:
Let's see how long we'll need to support intel macs. Once apple drops rosetta, I think we'll stop.

angerman reviewed (changes requested):
Looks great! Just a few minor nits.

hasufell committed:
Don't build ghcjs on mac x86_64

hasufell commented:
@copilot you're useless

hasufell committed:
Filter ld garbage on mac

hasufell committed:
Revert "rts: Add stgCallocAlignedBytes with zeroing semantics"

This reverts commit 5296ed7b36427f80bc295c29dfa408d0d6384367.

hasufell commented:
Pushed to `stable-ghc-9.14`
